token-lending/program/src/state/obligation.rs

```diff
@@ -1,186 +1,223 @@
+    pub loan_to_value_ratio: Rate,
+    /// The loan to value ratio at which the obligation can be liquidated
+    pub liquidation_threshold: Rate,
 }
 
 impl Obligation {
-    /// Create new obligation
-    pub fn new(params: NewObligationParams) -> Self {
-        let NewObligationParams {
-            collateral_reserve,
-            borrow_reserve,
-            token_mint,
-            cumulative_borrow_rate_wads,
-        } = params;
-
-        Self {
-            version: PROGRAM_VERSION,
-            deposited_collateral_tokens: 0,
-            collateral_reserve,
-            cumulative_borrow_rate_wads,
-            borrowed_liquidity_wads: Decimal::zero(),
-            borrow_reserve,
-            token_mint,
-        }
+    /// Create a new obligation
+    pub fn new(params: InitObligationParams) -> Self {
+        let mut obligation = Self::default();
+        Self::init(&mut obligation, params);
+        obligation
     }
 
-    /// Maximum amount of loan that can be closed out by a liquidator due
-    /// to the remaining balance being too small to be liquidated normally.
-    pub fn max_closeable_amount(&self) -> Result<u64, ProgramError> {
-        if self.borrowed_liquidity_wads < Decimal::from(CLOSEABLE_AMOUNT) {
-            self.borrowed_liquidity_wads.try_ceil_u64()
-        } else {
-            Ok(0)
-        }
+    /// Initialize an obligation
+    pub fn init(&mut self, params: InitObligationParams) {
+        self.version = PROGRAM_VERSION;
+        self.last_update = LastUpdate::new(params.current_slot);
+        self.lending_market = params.lending_market;
+        self.owner = params.owner;
+        self.deposits = params.deposits;
+        self.borrows = params.borrows;
     }
 
-    /// Maximum amount of loan that can be repaid by liquidators
-    pub fn max_liquidation_amount(&self) -> Result<u64, ProgramError> {
-        Ok(self
-            .borrowed_liquidity_wads
-            .try_mul(Rate::from_percent(LIQUIDATION_CLOSE_FACTOR))?
-            .try_floor_u64()?)
-    }
-
-    /// Ratio of loan balance to collateral value
-    pub fn loan_to_value(
+    /// Calculate the maximum liquidation amount for a given liquidity
+    pub fn max_liquidation_amount(
         &self,
-        collateral_exchange_rate: CollateralExchangeRate,
-        borrow_token_price: Decimal,
+        liquidity: &ObligationLiquidity,
     ) -> Result<Decimal, ProgramError> {
-        let loan = self.borrowed_liquidity_wads;
-        let collateral_value = collateral_exchange_rate
-            .decimal_collateral_to_liquidity(self.deposited_collateral_tokens.into())?
-            .try_div(borrow_token_price)?;
-        loan.try_div(collateral_value)
+        let max_liquidation_value = self
+            .borrowed_value
+            .try_mul(Rate::from_percent(LIQUIDATION_CLOSE_FACTOR))?
+            .min(liquidity.market_value);
+        let max_liquidation_pct = max_liquidation_value.try_div(liquidity.market_value)?;
```

Comment:
Possible that `market_value` is zero here? Might want a nicer error if so

Reply:
Could borrowed value be a small number and market value get rounded to zero?

Reply:
I'll think about that more and test out the smallest valid values for `market_price` and `borrowed_amount_wads`. If it's smaller than `LIQUIDATION_CLOSE_AMOUNT = 2` it won't hit this code, but that doesn't totally answer the question.

Reply:
What about checking `market_price` on reserve refresh and `market_value` on obligation refresh to be nonzero and error otherwise? This would make it so that if they are zero, these reserves and obligations can't be refreshed, which is kind of a safety over liveliness guarantee.

Reply:
So that would mean that if a token went to near zero value, no one would be able to withdraw?

Reply:
When you say it like that it sounds less smart 😅
But I think it would only be true if the token price went to almost zero AND they had almost zero tokens deposited, right? Because the market price being near zero on reserve's refreshing isn't inherently a problem, but it's a problem when the obligation's deposited/borrowed amount is multiplied and that ends up being less than 1e-18.

Reply:
But yeah, either way I don't think it makes sense to halt withdraws if that happens I suppose, though I think you could end up in a weird situation if `borrowed_value` gets close to zero but there are still borrows and you withdraw collateral.

Reply:
Liquidity and collateral market value is now checked to be nonzero in the liquidation processor function @ c1067fc63cdb0d0b3af47c193dc8f3e6fd04373a